### ISSUE TITLE
Upgrade to RN v0.60

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3548]
Released last week:
https://facebook.github.io/react-native/blog/2019/07/03/version-60

~First we'll want to complete the upgrade to v0.59 `<_<` `>_>`, aka #3399.~ (done!)

From the release notes, notable changes include:
* Migrated to AndroidX -- this will take some work on upgrading.
* [edit:] CocoaPods. See #3983.
* There's a new recommended upgrade process.
* (I should read the rest of the release notes more completely -- getting this down quickly for now.)

Also:
* Flow v0.98.

---

The bulk of this work is now done. Details in https://github.com/zulip/zulip-mobile/issues/3548#issuecomment-638372833 and in #4133 .

Some known changes that we already take advantage of in #4133:
* We can start using "autolinking": #4026
* We upgrade react-native-webview, fixing in particular #4093.

Other known changes that this will open up the option to do -- in particular I went through this thread and followed all the backlinks to related issues:
* We can start trying Hermes: #4131
* We can probably upgrade react-native-unimodules: #4091
  * That may also let us drop a bit of Gradle code: https://github.com/zulip/zulip-mobile/issues/3548#issuecomment-607558197
* We'll gain the feature we need for solving #2772.
* The react-navigation v3 upgrade #3649 becomes a bit simpler.
* We'll gain support for Android's "Downloadable Fonts" feature, which will give us what we need to fix #3885.
* We'll gain a better way to help the user to the permissions menu on iOS: #3814

Known cleanups we should probably follow this with:
* #4118 react-native-orientation

**Comment by gnprice** [comment]
#3399 is now done! We're on RN v0.59, and so this upgrade to v0.60 is no longer blocked.

**Comment by ray-kraesig** [comment]
A list of known dependencies whose update is blocked by the transition to RN v0.60:
* react-native-image-picker
* *(add more as identified)*

**Comment by gnprice** [comment]
(Belatedly updating the assignment metadata.)

**Comment by gnprice** [comment]
Good news on the AndroidX upgrade! I sat down yesterday afternoon to try it, and it looks like it will be (a) a small change that (b) does not depend on the RN upgrade, and that in fact we could have done at any time in the last year or so.

I have a draft commit to do this; I think it's most likely complete as is, but there are a few more things I want to test.

Detailed discussion in this (long!) commit message: 5cec78fe4

**Comment by chrisbobbe** [comment]
From my recent work on #3987 and #4016, here's some work that we'd like to get done as part of this upgrade:

- This is already pointed out in the [Upgrade Helper tool](https://react-native-community.github.io/upgrade-helper/), but we'll need to change our Podfile so it looks more like [this](https://raw.githubusercontent.com/react-native-community/rn-diff-purge/release/0.60.0/RnDiffApp/ios/Podfile) because of https://github.com/facebook/react-native/commit/2321b3fd7, which was released in RN v0.60.0.

**Comment by gnprice** [comment]
Returning to this issue because #4091 is a reminder that it'd be great to be on a more recent version.

I believe at this point we've largely completed the hard work required for this:
* adjusting to the newer Flow version (#3827)
* updating to AndroidX (#3852, which I should rebase and merge)
* migrating to CocoaPods (#3987)

The remaining tasks are
* Rebase, re-test, and merge that AndroidX upgrade #3852 -- in particular with attention to this point from cb87f9062:

     In that same comment, a maintainer says that 0.7.0-rc.4 is the
    minimum version with AndroidX compatibility. If this is true, then
    we may have a slight problem, because 0.7.0-rc.4, just like 0.6.0,
    does not work with RN 0.59.10. That would mean we had to do the
    AndroidX upgrade (PR #3852) at the same time as the RN v0.60 upgrade
    (issue #3548). Greg thinks it's likely that the maintainer is
    mistaken that there's a minimum version of unimodules required to
    use AndroidX; the AndroidX engineers designed a smooth migration
    path such that this shouldn't happen (see Greg's comment at
    https://github.com/zulip/zulip-mobile/pull/3987#issuecomment-601918696).

* Try the upgrade, test for any remaining breakage, and triage that

* Fix that remaining breakage :slightly_smiling_face: which hopefully is all small

**Comment by chrisbobbe** [comment]
(As discussed around [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/apple.20purpose.20strings/near/869947), I think that paragraph from my commit message for cb87f90 is correct if you remove "just like 0.6.0". I'm not sure how that made it in, but it's not true; we're happily using 0.6.0 with RN v0.59.10. My apologies.)

I also think most of the work of the RN v0.60 upgrade has been done. @gnprice, I'm happy to drive that, if you like (following your merge of #3852), but I want to finish Sign in with Apple first, if that's OK. I think it's wise to suppose a RN upgrade will be full of nice surprises, some of which could lurk undetected in our dependencies for days or weeks after a main PR is merged, and will have to be dealt with, possibly by the person who did the actual upgrade. (Obviously we'll try our best to avoid this happening.)

**Comment by gnprice** [comment]
That plan sgtm!

**Comment by chrisbobbe** [comment]
Following `facebook/react-native@2321b3fd7`, we're pretty sure it's fine for any dependency's `.podspec` file to have a line like `s.dependency "React"` when it depends on code specific to one of the modules that was previously a "subspec" but is now a separate pod, e.g., "React/Core" (the "Core" subspec) n.k.a. "React-Core" (a pod thus named). The "React/Core" / "React-Core" is, AFAIK, the only one that might need to be examined, because a `default_subspec` was declared to be "Core" in "React.podspec", previously, so "React/Core" could have been spelled "React", while, e.g., "React/RCTImage" could not have been spelled "React". More speculation, experiments, and links [here](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23M3548.20RN.20v0.2E60.2E0.20upgrade/near/880638).

**Comment by chrisbobbe** [comment]
As Greg suggests at https://github.com/zulip/zulip-mobile/pull/4133#issuecomment-637928734, I'm cutting out the parts from the commit message that talk about work *not* done in the upgrade series, and putting it here:

## Platform-agnostic

- We've left "scripts" in `package.json` alone; ours work fine as-is.

## Android

- Adding the `debug.keystore` file and its config in
  `android/app/build.gradle`; we don't take these changes at all.
  See 9ccaa2125.

- Adding `android.useAndroidX=true` and
  `android.enableJetifier=true` in `android/gradle.properties`.
  These were done in e433197cf.

- Deleting `android/keystores/BUCK` and
  `android/keystores/debug.keystore.properties`: We don't use Buck
  (removal of some of its boilerplate was 1c86488ce), and we don't
  have an `android/keystores` directory. As mentioned in 9ccaa2125,
  we're happy with the Android Studio-provided debug keystore, and
  our own release keystore described in our release guide.

## iOS

Some iOS changes from the upgrade guide don't appear in this series.
They fall neatly into a few touched files:

- ios/ZulipMobile-tvOS/Info.plist: We're not (yet) in the TV
  business. ;)

- ios/ZulipMobile.xcodeproj/project.pbxproj: The purge of many large
  chunks of this file, often known as "the Xcode config file", was
  already done, in 33f4b41c8.

- ios/ZulipMobile.xcworkspace/contents.xcworkspacedata: After
  33f4b41c8, this file already looks the way it should.

- ios/ZulipMobile/Info.plist: These changes are done upstream in
  facebook/react-native@7b44fe56f. They fix a duplication issue and
  a code-comment issue, which were apparently causing build
  failures. Our code doesn't have these issues to begin with.
